### PR TITLE
Hydjet integration as external

### PR DESCRIPTION
**Pull request by wouf**
Hydjet 1.9.1 integration as external. Update pyquen.

PR cmssw: https://github.com/cms-sw/cmssw/pull/30185

### CODE REVIEW

**Comment by smuzaffar** [comment]
assign generators

File: hydjet.spec [hunk 1]
```diff
@@ -0,0 +1,20 @@
+### RPM external hydjet 1.9.1
+
+Source: http://cern.ch/lokhtin/hydro/%{n}-%{realversion}.tar.gz
+
+BuildRequires: cmake
```
Comment (smuzaffar) [inline hydjet.spec]: please also add build dependency on `gmake` otherwise it will pick system gmake.

File: hydjet.spec [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+%setup -q -n %{n}-%{realversion}
+
+%build
+
+cmake . -DCMAKE_INSTALL_PREFIX=%i -DCMAKE_BUILD_TYPE=Release -DPYQUEN_DIR=${PYQUEN_ROOT} -DPYTHIA6_DIR=${PYTHIA6_ROOT} -DLHAPDF_ROOT_DIR=${LHAPDF_ROOT}
+cmake --build . --clean-first -- %{makeprocesses}
+
+%install
+
+cmake --build . --target install --clean-first -- %{makeprocesses}
```
Comment (smuzaffar) [inline hydjet.spec]: please remove `--clean-first` here otherwise it will clean and re-build

**Comment by smuzaffar** [comment]
in order to get this in cmssw environment please update `cmssw-tool-conf.spec` and add dependency on `hydjet-toolfile`

File: hydjet.spec (hunk 1 above)
Comment (wouf) [reply hydjet.spec]: Thanks for advice. But, does it matter?

File: hydjet.spec (hunk 2 above)
Comment (wouf) [reply hydjet.spec]: Yes, that was the goal (to clean in case the previous build found). What's wrong? The same behaviour was realized for pyquen.

**Comment by wouf** [comment]
Thanks, I added hydjet-toolfile to cmssw-tool-conf.spec in patch-7

File: hydjet.spec (hunk 2 above)
Comment (smuzaffar) [reply hydjet.spec]: there is no need to do it twice. It has been done in the %build section already

Comment (wouf) [reply hydjet.spec]: I see, thanks!

File: hydjet.spec (hunk 1 above)
Comment (smuzaffar) [reply hydjet.spec]: yes it does, some times we end up in env (e.g. docker containers with minimal package installation) where gmake is not available via system

Comment (wouf) [reply hydjet.spec]: I mean does it matter gmake, make, or ninja tool is used?

Comment (smuzaffar) [reply hydjet.spec]: it depends,  if you have run cmake to use ninja then  you need to add dependency on ninja otheriwse default is `make` (which is available via our gmake.spec) and for that you need to add dependency on gmake.

Comment (wouf) [reply hydjet.spec]: Thank You for Your explanation!

**Comment by smuzaffar** [comment]
please test workflow 300.0

**Comment by fabiocos** [comment]
@wouf @hqucms @agrohsje I understand that this PR moves hydjet, currently within CMSSW, outside as an external. I imagine therefore that there should be a corresponding update of https://cmssdt.cern.ch/lxr/source/GeneratorInterface/HydjetInterface/ please clarify.

Is there a validation of the correctness of this backport wrt the original implementation?

**Comment by wouf** [comment]
@fabiocos sure we tested it with the new GeneratorInterface/HydjetInterface/. I was going to update it soon, but I realised that new (as well as existing, please see https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/HydjetInterface/src/HydjetHadronizer.cc#L195) interface has an issue: we trying to read empty edm::event using BaseHadroniser::getEDMEvent function. To correct it we have to call BaseHadroniser::setEDMEvent somewhere, but I can't to figure out how to access edm::event (from HiMixGEN) in GeneratorInterface. We need to fix it in the new HydjetInterface. Do You have any ideas on this? Example of workflow which is needs it: 

`cmsDriver.py Hydjet_Quenched_MinBias_5020GeV_cfi --conditions auto:phase1_2018_realistic --scenario HeavyIons -n 1 --era Run2_2018 --eventcontent RAWSIM --relval 2000,1 -s GEN,SIM --datatier GEN-SIM --beamspot Realistic25ns13TeVEarly2018Collision --geometry DB:Extended --pileup HiMixGEN --pileup_input file:Hydjet_Quenched_B12_5020GeV_cfi_GEN_SIM.root --fileout step1_GEN_SIM_PU.root --no_exec`

The corresponding issue #28636 was created.

**Comment by silviodonato** [comment]
@wouf do you have any news on this PR?

**Comment by wouf** [comment]
I added the interface but we still working on [issue 28636](https://github.com/cms-sw/cmssw/issues/28636)

**Comment by qliphy** [comment]
Andrey gave a talk at Generator meeting with local tests look good.
https://indico.cern.ch/event/932445/contributions/3935235/attachments/2073197/3480861/HJ1.pdf

**Comment by qliphy** [comment]
+1

**Comment by mrodozov** [comment]
please test
@wouf could you please provide description and documentation (if possible) in the hydjet/doc.txt file like in here:
https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_11_2_X/master/fmt/doc.txt
we want to keep track who added what, why and some basic description of the external.
besides that this one can go in (from externals perspective)

**Comment by mrodozov** [comment]
please test

**Comment by wouf** [comment]
@mrodozov I added for pyquen as well

**Comment by mrodozov** [comment]
please test

**Comment by qliphy** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Pull request merged**